lib/src/entrypoint.dart

```diff
@@ -1335,13 +1335,27 @@ See https://dart.dev/go/sdk-constraint
   /// Remove any `pubspec.lock` or `.dart_tool/package_config.json` files in
   /// workspace packages that are not the root package.
   ///
+  /// Also remove from directories between the workspace package and the
+  /// workspace root, to prevent stray package configs from shadowing the shared
+  /// workspace package config.
+  ///
   /// This is to avoid surprises if a package is turned into a workspace member
   /// but still has an old package config or lockfile.
   void _removeStrayLockAndConfigFiles() {
+    final visited = <String>{};
     for (final package in workspaceRoot.transitiveWorkspace) {
       if (package.pubspec.resolution == Resolution.workspace) {
-        deleteEntry(p.join(package.dir, 'pubspec.lock'));
-        deleteEntry(p.join(package.dir, '.dart_tool', 'package_config.json'));
+        for (final dir in parentDirs(package.dir)) {
+          if (p.equals(dir, workspaceRoot.dir)) {
+            break;
+          }
+          if (!visited.add(dir)) {
+            // No reason to delete from the same directory twice.
+            break;
+          }
+          deleteEntry(p.join(dir, 'pubspec.lock'));
+          deleteEntry(p.join(dir, '.dart_tool', 'package_config.json'));
```

Comment:
Should we write a warning message or something to logs.

Maybe something that includes a dart.dev/go/-link, so that we can explain why we deleted these folders.

If someone has workspace packages nested inside non-workspace packages inside a root, that could cause this, right? or probably it'll cause an error when trying to resolve the workspaces packages?

Reply:
> Should we write a warning message or something to logs.

I'm not completely against this, but I think it might confuse more than it benefits...

> Maybe something that includes a dart.dev/go/-link,

I think that would be slightly overdoing it... But maybe we can fit a paragraph in somewhere on the workspaces documentation...

> so that we can explain why we deleted these folders.

We are deleting files, not folders.

> If someone has workspace packages nested inside non-workspace packages inside a root,

We should probably detect if someone attempts this (by detecting a pubspec.yaml), and give an error message. 
That scenario is not supported.

Reply:
Ok - now displaying a warning and a doc reference.

Also added an error check if there are stray pubspec.yaml files between a package and the root.